src/bvh/ordered_traverse.rs

```diff
@@ -0,0 +1,345 @@
+    ///
+    /// # Panics
+    ///
+    /// Panics if `stack_size` underflows.
+    fn stack_pop(&mut self) -> (usize, RestChild) {
+        self.stack_size -= 1;
+        self.stack[self.stack_size]
+    }
+
+    /// Attempt to move to the child node that is closest to the ray, relative to the current node.
+    /// If it is a leaf, or the [`Ray`] does not intersect the any node [`Aabb`], `has_node` will become `false`.
+    fn move_nearest(&mut self) -> (usize, RestChild) {
+        let current_node_index = self.node_index;
+        match self.bvh.nodes[current_node_index] {
+            BvhNode::Node {
+                child_l_index,
+                ref child_l_aabb,
+                child_r_index,
+                ref child_r_aabb,
+                ..
+            } => {
+                let (left_dist, _) = self.ray.intersection_slice_for_aabb(child_l_aabb);
+                let (right_dist, _) = self.ray.intersection_slice_for_aabb(child_r_aabb);
+
+                if left_dist < T::zero() {
+                    if right_dist < T::zero() {
+                        // no intersections with any children
+                        self.has_node = false;
+                        (current_node_index, RestChild::None)
+                    } else {
+                        // has intersection only with right child
+                        self.has_node = true;
+                        self.node_index = child_r_index;
+                        let rest_child = RestChild::None;
+                        (current_node_index, rest_child)
+                    }
+                } else if right_dist < T::zero() {
+                    // has intersection only with left child
+                    self.has_node = true;
+                    self.node_index = child_l_index;
+                    let rest_child = RestChild::None;
+                    (current_node_index, rest_child)
+                } else if left_dist > right_dist {
+                    // right is closer than left
+                    self.has_node = true;
+                    self.node_index = child_r_index;
+                    let rest_child = RestChild::Left;
+                    return (current_node_index, rest_child);
+                } else {
+                    // left is closer than right
+                    self.has_node = true;
+                    self.node_index = child_l_index;
+                    let rest_child = RestChild::Right;
+                    return (current_node_index, rest_child);
+                }
+            }
+            BvhNode::Leaf { .. } => {
+                self.has_node = false;
+                (current_node_index, RestChild::None)
+            }
+        }
+    }
+
+    /// Attempt to move to the child node that is the farthest from the ray, relative to the current node.
+    /// If it is a leaf, or the [`Ray`] does not intersect the node [`Aabb`], `has_node` will become `false`.
+    fn move_furthest(&mut self) -> (usize, RestChild) {
```

Comment:
I see that 95% of the code between move_furthest and move_nearest is duplicated, other than line 149 and 150 which choose the farthest intersection slice, and like 170 which has the condition reversed. It seems like these could be made to pick one depending on a generic Boolean.

Would probably make the code easier to maintain in the future.